platform/drm/cog-drm-gles-renderer.c

```diff
@@ -0,0 +1,567 @@
+    glBindTexture(GL_TEXTURE_2D, self->gl_texture);
+    glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, wpe_fdo_egl_exported_image_get_egl_image(image));
+    glUniform1i(self->gl_uniform_texture, 0);
+
+    /* clang-format off */
+    static const float position_coords[4][2] = {
+        { -1,  1 }, { 1,  1 },
+        { -1, -1 }, { 1, -1 },
+    };
+    static const float texture_coords[4][4][2] = {
+        [COG_DRM_RENDERER_ROTATION_0] = {
+            { 0, 0 }, { 1, 0 },
+            { 0, 1 }, { 1, 1 },
+        },
+        [COG_DRM_RENDERER_ROTATION_90] = {
+            { 1, 0 }, { 1, 1 },
+            { 0, 0 }, { 0, 1 },
+        },
+        [COG_DRM_RENDERER_ROTATION_180] = {
+            { 1, 1 }, { 0, 1 },
+            { 1, 0 }, { 0, 0 },
+        },
+        [COG_DRM_RENDERER_ROTATION_270] = {
+            { 0, 1 }, { 0, 0 },
+            { 1, 1 }, { 1, 0 },
+        },
+    };
+    /* clang-format on */
+
+    glVertexAttribPointer(self->gl_attrib_position, 2, GL_FLOAT, GL_FALSE, 0, position_coords);
+    glVertexAttribPointer(self->gl_attrib_texture, 2, GL_FLOAT, GL_FALSE, 0, texture_coords[self->rotation]);
+
+    glEnableVertexAttribArray(self->gl_attrib_position);
+    glEnableVertexAttribArray(self->gl_attrib_texture);
+
+    glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
+
+    glDisableVertexAttribArray(self->gl_attrib_position);
+    glDisableVertexAttribArray(self->gl_attrib_texture);
+
+    if (G_UNLIKELY(!eglSwapBuffers(self->egl_display, self->egl_surface))) {
+        g_critical("%s: eglSwapBuffers failed (%#04x)", __func__, eglGetError());
+        return;
+    }
+
+    wpe_view_backend_exportable_fdo_egl_dispatch_release_exported_image(self->exportable, image);
+
+    int drm_fd = gbm_device_get_fd(self->gbm_device);
+
+    struct gbm_bo *bo = gbm_surface_lock_front_buffer(self->gbm_surface);
+
+    uint32_t handles[4], strides[4], offsets[4];
+    uint64_t modifiers[4];
+
+    for (unsigned i = 0; i < gbm_bo_get_plane_count(bo); i++) {
+        handles[i] = gbm_bo_get_handle_for_plane(bo, i).u32;
+        strides[i] = gbm_bo_get_stride_for_plane(bo, i);
+        offsets[i] = gbm_bo_get_offset(bo, i);
+        modifiers[i] = gbm_bo_get_modifier(bo);
+    }
+
+    uint32_t fb_id = 0;
+    uint32_t flags = (modifiers[0] && modifiers[0] != DRM_FORMAT_MOD_INVALID) ? DRM_MODE_FB_MODIFIERS : 0;
+    int ret = drmModeAddFB2WithModifiers(drm_fd, self->mode.hdisplay, self->mode.vdisplay, self->gbm_format, handles,
+                                         strides, offsets, modifiers, &fb_id, flags);
+    if (ret) {
+        handles[0] = gbm_bo_get_handle(bo).u32;
+        handles[1] = handles[2] = handles[3] = 0;
+        strides[0] = gbm_bo_get_stride(bo);
+        strides[1] = strides[2] = strides[3] = 0;
+        offsets[0] = offsets[1] = offsets[2] = offsets[3] = 0;
+        ret = drmModeAddFB2(drm_fd, self->mode.hdisplay, self->mode.vdisplay, self->gbm_format, handles, strides,
+                            offsets, &fb_id, 0);
+    }
```

Comment:
Okay, I have been reading around and the “retry without modifiers on failure” dance (which I cargo-culted from the existing DRM code, maybe itself cargo-culted from elsewhere) is needed because BO creation typically does _not_ query the KMS subsystem for appropriate/needed modifiers, but at some point the ultimate responsible for allocating the BO could have picked some modifiers which may or may not be supported by the KMS plane used for presenting the framebuffer. Retrying without the modifiers can get things working in some cases and I have seen it as a recommendation [e.g. in the LKML](https://lkml.org/lkml/2020/7/1/1118)—to be done _even_ if `DRM_CAP_ADDFB2_MODIFIERS` is supported by the DRM/KMS output.

I'll add a comment here about this.